File: app/backend/src/api/routes/manager.py
from __future__ import annotations

from fastapi import APIRouter

from api.errors import ApiError
from features.models import service as models
from features.models.schemas import ManagerInstallIn

api = APIRouter()


@api.get("/manager/models")
def get_manager_models(mode: str = "cache") -> dict:
    try:
        return models.list_manager_models(mode)
    except models.ManagerCatalogError as exc:
        raise ApiError(exc.code, str(exc), exc.status) from exc


@api.post("/manager/models/install")
def post_manager_install(body: ManagerInstallIn) -> dict:
    try:
        return models.install_manager_model(body.name, body.filename, body.save_path)
    except models.ManagerCatalogError as exc:
        raise ApiError(exc.code, str(exc), exc.status) from exc

File: app/frontend/src/lib/api/manager.ts
import { api, readError } from './http.ts'

export type ManagerModel = {
  name: string
  type: string
  base: string
  save_path: string
  description: string
  reference: string
  filename: string
  size: string
  installed: 'True' | 'False'
}

export async function listManagerModels(mode = 'cache'): Promise<ManagerModel[]> {
  const res = await fetch(api(`/manager/models?mode=${encodeURIComponent(mode)}`))
  if (!res.ok) {
    throw new Error(await readError(res))
  }
  const data = (await res.json()) as { models?: ManagerModel[] }
  return Array.isArray(data.models) ? data.models : []
}

export async function installManagerModel(item: Pick<ManagerModel, 'name' | 'filename' | 'save_path'>): Promise<void> {
  const res = await fetch(api('/manager/models/install'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name: item.name, filename: item.filename, save_path: item.save_path }),
  })
  if (!res.ok) {
    throw new Error(await readError(res))
  }
}
